pyds9/pyds9.py

```diff
@@ -610,184 +630,270 @@ def access(self):
-            important in the case where the array has datatype np.uint64,
-            which is not recognized by ds9::
-
-                >>> d.set_np2arr(arru64)
-                ...
-                ValueError: uint64 is unsupported by DS9 (or FITS)
-                >>> d.set_np2arr(arru64,dtype=np.float64)
-                1
-
-            Also note that np.int8 is sent to ds9 as int16 data, np.uint32 is
-            sent as int64 data, and np.float16 is sent as float32 data.
-            """
-            self._selftest()
-            if type(arr) != numpy.ndarray:
-                raise ValueError('requires numpy.ndarray as input')
-            if dtype and dtype != arr.dtype:
-                narr = arr.astype(dtype)
-            else:
-                if arr.dtype == numpy.int8:
-                    narr = arr.astype(numpy.int16)
-                elif arr.dtype == numpy.uint32:
-                    narr = arr.astype(numpy.int64)
-                elif hasattr(numpy, "float16") and arr.dtype == numpy.float16:
-                    narr = arr.astype(numpy.float32)
-                else:
-                    narr = arr
-            if not narr.flags['C_CONTIGUOUS']:
-                narr = numpy.ascontiguousarray(narr)
-            bp = _np2bp(narr.dtype)
-            buf = narr.tostring('C')
-            blen = len(buf)
-            (w, h) = narr.shape
-
-            # note that this needs the "endian=" part because sometimes it's
-            # left out completely
-            endianness = ''
-            if narr.dtype.byteorder == '=':
-                endianness = ',endian=' + sys.byteorder
-            elif narr.dtype.byteorder == '<':
-                endianness = ',endian=little'
-            elif narr.dtype.byteorder == '>':
-                endianness = ',endian=big'
-
-            paramlist = 'array [xdim={0},ydim={1},bitpix={2}{3}]'
-            return self.set(paramlist.format(h, w, bp, endianness), buf,
-                            blen+1)
+        Returns
+        -------
+        :class:`six.BytesIO`
+            file-like object containing the FITS data
+        '''
+        self._selftest()
+        imgData = self.get('fits')
+        return BytesIO(string_to_bytes(imgData))
+
+    def _hdulist_to_ds9_fits(self, hdul):
+        '''Send the input HDUList to ds9
+
+        Parameters
+        ----------
+        hdul : :class:`HDUList` (astropy or pyfits)
+            FITS file to send
+
+        Returns
+        -------
+        success : int
+            1 indicates that ds9 was contacted successfully, while a return
+            value of 0 indicates a failure.
+        '''
+        self._selftest()
+        # for python2 BytesIO and StringIO are the same
+        with contextlib.closing(BytesIO()) as newFitsFile:
+            hdul.writeto(newFitsFile)
+            newfits = newFitsFile.getvalue()
+            success = self.set('fits', newfits, len(newfits))
+        return success
+
+    def get_fits(self):
```

Comment:
get a fits from ds9 as an ``astropy`` HDUList

Reply:
I just realised that I need to thing about backward compatibility. I'll think something